### PR TITLE
feat(pb): box stream NodeBody to reduce stack memory usage

### LINKED ISSUE #19910: frontend stack overflow at stream fragmenter
To reproduce: TPCH q5

It's not stack depth too deep, but memory usage too large. So #16279 doesn't help.
(On mac, default stack size is 8MB.)

Analysis:
- stream `NodeBody`: 4120 bytes. (BTW, batch node body is 680 bytes)
    - largest is `HashJoin` 4120 bytes, 2nd largest is DyanamicFilter` 2144 bytes. 
    - small ones: `Project` 112 bytes, `Filter` 336 bytes
    - It's an unbalanced `enum`, so we will use 4120 bytes for all nodes.
- The main reason of the large plan node is `Table`: 904 bytes (`HashJoin` has 4 tables)
    - Largest problem: recently introduced `webhook_info` is 344 bytes. (That's why we hit the stackoverflow problem recently but not before)
        - This is because `ExprNode` is 328 bytes, where `DataType` is 72 bytes and `RexNode` is 248 bytes.
        - For `RexNode`, `Udf` is 248 bytes (!), and other branch is 32 bytes.
    - Some other minor issues: `Option<u32>` wasted 4 bytes, since the zero value is not used, or we don't need this large range. We may use `i32` or `u32` instead.
    - 41 fields in total! @st1page said that this was a mistake: we should pub sth like `TableDesc` in plan, instead of `TableCatalog`.

Mitigate:
- make `expr.ExprNode.rex_node.udf` boxed: `ExprNode` becomes 112 bytes, `Table` 688 bytes, `NodeBody` 3040 bytes. So this seems not enough.
- Currently the easiest way seems to be just boxing all variants of `NodeBody`. I'm not sure whether this is optimal.
- Alternatively we can box all occurences of `Table`

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://raw.githubusercontent.com/risingwavelabs/risingwave/17af8a747593ebdbfa826691daf75bdab7d14fa0/.github/contributor-license-agreement.txt).

## What's changed and what's your intention?
fix #19910 See analysis in the issue

also revert #19695

## Checklist

- [ ] I have written necessary rustdoc comments.
- [ ]  I have added necessary unit tests and integration tests.
- [ ]  I have added test labels as necessary. 
- [ ]  I have added fuzzing tests or opened an issue to track them. 
- [ ]  My PR contains breaking changes. 
- [ ]  My PR changes performance-critical code, so I will run (micro) benchmarks and present the results. 
- [ ]  My PR contains critical fixes that are necessary to be merged into the latest release. 

## Documentation

- [ ]  My PR needs documentation updates. 

<details>
<summary><b>Release note</b></summary>

</details>

### CODE REVIEW

File: src/prost/src/lib.rs [hunk 1]
```diff
@@ -551,4 +560,11 @@ mod tests {
         };
         assert!(!new_data_type.is_nullable);
     }
+
+    #[test]
+    fn test_size() {
+        use static_assertions::const_assert_eq;
+
+        const_assert_eq!(std::mem::size_of::<NodeBody>(), 16);
```
Comment (xxchan) [inline src/prost/src/lib.rs:568]: 4120 bytes -> 16 bytes

File: src/prost/build.rs [hunk 2]
```diff
@@ -112,6 +112,57 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
             "#[derive(::enum_as_inner::EnumAsInner)]",
         )
         .btree_map(btree_map_paths)
+        // node body is a very large enum, so we box it to avoid stack overflow.
+        // TODO: ideally we should box all enum variants automatically https://github.com/tokio-rs/prost/issues/1209
+        .boxed(".stream_plan.StreamNode.node_body.source")
```
Comment (fuyufjh) [inline src/prost/build.rs:117]: Why not `.boxed(".stream_plan.StreamNode")`?

**Review by fuyufjh — APPROVED** [review]

**Review by stdrc — APPROVED** [review]
LGTM

**Review by MrCroxx — APPROVED** [review]
LGTM

**Review by st1page — APPROVED** [review]

File: src/prost/build.rs [hunk 3]
```diff
@@ -112,6 +112,57 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+        .boxed(".stream_plan.StreamNode.node_body.project")
+        .boxed(".stream_plan.StreamNode.node_body.filter")
+        .boxed(".stream_plan.StreamNode.node_body.materialize")
```
Comment (chenzl25) [inline src/prost/build.rs:120]: Should we box the batch plan node as well?

File: src/prost/build.rs (hunk 2 above)
Comment (xxchan) [reply src/prost/build.rs:117]: Sounds can also fix the stackoverflow. 🤔 But the enum itself still wastes memory.

File: src/prost/build.rs (hunk 3 above)
Comment (xxchan) [reply src/prost/build.rs:120]: batch node body was 680 bytes. After boxing Udf in Expr, it's 464 bytes. Sounds acceptable

Comment (xxchan) [reply src/prost/build.rs:120]: LocalLookupJoinNode is the largest

File: src/prost/build.rs (hunk 2 above)
Comment (BugenZhao) [reply src/prost/build.rs:117]: Listing all possible variants here seems verbose. Is this the only approach to achieve the goal?

File: src/prost/src/lib.rs [hunk 4]
```diff
@@ -551,4 +562,11 @@ mod tests {
...
```
Comment (BugenZhao) [inline src/prost/src/lib.rs:570]: May add some comments to instruct future developers to box the newly introduced variant if this assertion fails.

Also, can use `static_assertions` from the standard library and there's no need to put it inside a `#[test]` function.

**Review by BugenZhao — APPROVED** [review]

File: src/prost/build.rs (hunk 2 above)
Comment (xxchan) [reply src/prost/build.rs:117]: can implement this upstream https://github.com/tokio-rs/prost/issues/1209

**Review by chenzl25 — APPROVED** [review]

**Review by Li0k — APPROVED** [review]

File: src/prost/src/lib.rs (hunk 4 above)
Comment (xxchan) [reply src/prost/src/lib.rs:570]: Where is the static_assertions in the standard library ?

Comment (BugenZhao) [reply src/prost/src/lib.rs:571]: Oh, there's not. I confused it with `assert_matches`. 😄

**Pull request merged**